client/ethereum.go

```diff
@@ -339,3 +406,88 @@ func (e *EthereumClient) errorReason(b ethereum.ContractCaller, tx *types.Transa
 	}
 	return abi.UnpackRevert(res)
 }
+
+// EthereumTransactionConfirmer is an interface for allowing callbacks when the client receives a new header
+type EthereumTransactionConfirmer interface {
+	ReceiveHeader(header *types.Header) error
+	Wait() error
+}
+
+// TransactionConfirmer is an implementation of EthereumTransactionConfirmer that checks whether tx are confirmed
+type TransactionConfirmer struct {
+	minConfirmations int
+	confirmations    int
+	eth              *EthereumClient
+	txHash           common.Hash
+	doneChan         chan struct{}
+	context          context.Context
+	cancel           context.CancelFunc
+}
+
+// NewTransactionConfirmer returns a new instance of the transaction confirmer that waits for on-chain minimum
+// confirmations
+func NewTransactionConfirmer(eth *EthereumClient, txHash common.Hash, minConfirmations int) *TransactionConfirmer {
+	ctx, ctxCancel := context.WithTimeout(context.Background(), eth.Network.Config().Timeout)
+	tc := &TransactionConfirmer{
+		minConfirmations: minConfirmations,
+		confirmations:    0,
+		eth:              eth,
+		txHash:           txHash,
+		doneChan:         make(chan struct{}, 1),
+		context:          ctx,
+		cancel:           ctxCancel,
+	}
+	return tc
+}
+
+// ReceiveHeader the implementation of the EthereumTransactionConfirmer that receives each block header and checks
+// tx confirmation
+func (t *TransactionConfirmer) ReceiveHeader(header *types.Header) error {
+	block, err := t.eth.Client.BlockByNumber(context.Background(), header.Number)
+	if err != nil {
+		return err
+	}
+	confirmationLog := log.Debug().Str("Network", t.eth.Network.ID()).
+		Str("Block Hash", block.Hash().Hex()).
+		Str("Block Number", block.Number().String()).Str("Tx Hash", t.txHash.Hex()).
+		Int("Minimum Confirmations", t.minConfirmations)
+	isConfirmed, err := t.eth.isTxConfirmed(t.txHash)
+	if err != nil {
+		return err
+	} else if isConfirmed {
+		t.confirmations++
+	}
+	if t.confirmations == t.minConfirmations {
```

Comment:
Worth an `>=` I think, just in case.

Reply:
Did this deliberately as if it's `>=` it can show more confirmations than min confirmations until you call `Wait` which can be confusing. It's impossible for it to be greater than min without being equal.